folly/lang/SafeAssert.h

```diff
@@ -130,7 +130,7 @@ FOLLY_INLINE_VARIABLE constexpr safe_assert_msg_types_one_fn
     safe_assert_msg_types_one{}; // a function object to prevent extensions
 
 template <typename... A>
-safe_assert_msg_type_s<decltype(safe_assert_msg_types_one(A{}))::value...>
+safe_assert_msg_type_s<decltype(safe_assert_msg_types_one(std::declval<A>()))::value...>
```

Comment:
It looks like a shorter syntax is available.
```suggestion
safe_assert_msg_type_s<decltype(safe_assert_msg_types_one((A)A{}))::value...>
```

Reply:
@yfeldblum thanks for the idea!
But I prefer to use `std::declval` here, because it explicitly claims that this is unevaluated context.

Reply:
I assumed `declval` might be slower to compile, but I am not seeing that assumption holding in my tests.

Reply:
The use of `declval` triggered some bug or pathological behavior somewhere. I don't know what the problem is, but that's why this was reverted.

Reply:
@yfeldblum should I do something with it?

Reply:
I have merged this again but with `(A)`. Hopefully that will not trigger the same bug.